src/start/AuthScreen.js

```
@@ -1,24 +1,109 @@
+    SafariView.removeEventListener('onDismiss', this.safariViewDismissEvent);
+  }
+
   handleTypeSelect = (authType: string) => {
-    this.props.setAuthType(authType);
-    this.props.pushRoute(authType);
+    const { realm } = this.props;
+
+    if (authType === 'google') {
+      // Google OAuth flow
+      this.beginOAuthFlow(`${realm}/accounts/login/google/`);
+    } else if (authType === 'github') {
+      // Github OAuth flow
+      this.beginOAuthFlow(`${realm}/accounts/login/social/github/`);
+    } else {
+      // Password auth flow
+      this.props.setAuthType(authType);
+      this.props.pushRoute(authType);
+    }
+  }
+
+  beginOAuthFlow = async (url) => {
+    // Generate a one time pad (OTP) to send up with the request
+    // The server XORs the API key with this OTP in its response to protect
+    // against malicious apps registering an identical URI scheme and trying
+    // to intercept credentials
+    this.otp = await generateOtp();
+    SafariView.show({ url: `${url}?mobile_flow_otp=${this.otp}` });
+  };
+
+  endOAuthFlow = (event) => {
+    const { realm, loginSuccess } = this.props;
+
+    SafariView.dismiss();
+
+    const url = parseURL(event.url, true);
+
+    // The OAuth callback should have the following format:
+    // zulip://login?realm={}&email={}&otp_encrypted_api_key={}
+    if (url.host === 'login') {
+      const query = url.query;
+
+      // Check for errors
+      if (query.realm !== realm) {
+        // eslint-disable-next-line no-console
+        console.error('Zulip realm does not match request.');
+        return;
+      }
+
+      if (!this.otp) {
+        // eslint-disable-next-line no-console
+        console.error('No one time pad stored. This auth redirect may not have been requested.');
+        return;
+      }
+
+      if (!query.email || !query.otp_encrypted_api_key) {
+        // eslint-disable-next-line no-console
+        console.error('No credentials returned in server response.');
```

We may even want to show this to the user explicitly since `console.error` is swallowed in Release.